crates/dht/src/hash.rs

```diff
@@ -0,0 +1,443 @@
+//! also stored redundantly, say 10 times, this means that 20,000 nodes are needed for the data
+//! to be maximally distributed. With higher replication, say 50 times, and a hash factor of 20,
+//! over 200,000 nodes are needed for maximal distribution.
+//!
+//! Each hash partition manages a [PartitionedTime] structure that is responsible for managing the
+//! time slices for that hash partition. The interface of this module is largely responsible for
+//! delegating the updating of time slices to the inner time partitions. This ensures that all the
+//! time partitions are updated in lockstep. This makes reasoning about the space-time partitioning
+//! easier.
+//!
+//! This module must be informed about ops that have been stored. There is no active process here
+//! that can look for newly stored ops. When a batch of ops is stored, the [PartitionedHashes] must
+//! be informed and will split the ops into the right hash partition based on the location of the op.
+//! Ops are then pushed to the inner time partition for each hash partition. It is the time
+//! partitions that are responsible for updating the combined hash values.
+
+use crate::PartitionedTime;
+use kitsune2_api::{
+    ArcLiteral, DynOpStore, K2Error, K2Result, StoredOp, Timestamp,
+};
+use std::collections::HashMap;
+
+/// A partitioned hash structure.
+///
+/// Partitions the hash structure into a configurable number of partitions. Each partition is
+/// responsible for managing the time slices for that partition.
+#[derive(Debug)]
+pub struct PartitionedHashes {
+    size: u32,
+    partitioned_hashes: Vec<HashPartition>,
+}
+
+#[derive(Debug)]
+struct HashPartition {
+    _arc: ArcLiteral,
+    partitioned_time: PartitionedTime,
+}
+
+impl PartitionedHashes {
+    /// Create a new partitioned hash structure.
+    ///
+    /// The `hash_factor` determines the number of partitions to create. A value of 32 will create
+    /// a single partition. Each unit decrease from 32 will double the number of partitions.
+    ///
+    /// Each hash partition owns a [PartitionedTime] structure that is responsible for managing
+    /// the time slices for that hash partition. Other parameters to this function are used to
+    /// create the [PartitionedTime] structure.
+    pub async fn try_from_store(
+        hash_factor: u8,
+        time_factor: u8,
+        current_time: Timestamp,
+        store: DynOpStore,
+    ) -> K2Result<Self> {
+        if hash_factor > 32 {
+            return Err(K2Error::other("Hash factor must be 32 or lower"));
+        }
+
+        let (size, num_buckets) = if hash_factor == 32 {
+            (u32::MAX, 1)
+        } else {
+            let size = 1u32 << hash_factor;
+
+            // We will always be one bucket short because u32::MAX is not a power of two. It is one less
+            // than a power of two, so the last bucket is always one short.
+            (size, (u32::MAX / size) + 1)
```

Comment:
This will overflow with hash_factor = 0.
```suggestion
            let compensation = if size == 1 {
                0
            } else {
                1
            }
            (size, (u32::MAX / size) + compensation)
```

Reply:
You are correct. This will perform poorly long before reaching 0 though, so I'm just going to reject inputs that set the number of partitions too high. Well above 0

Reply:
This is now a fixed value